### ISSUE TITLE
Why is Javaparser not thread-safe?

### ISSUE THREAD

**Issue by corradomio** [javaparser/javaparser#3405]
I have to analyze 15.000/20.000 files.
The idea is to use Javaparser in parallel.
Then the problem is: why is javaparser not thread safe?
Which classes have problems?

I analyzed all "static" keywords present in the source code by I have not found the class that can have problems.
However this class there exists because using the current implementation in parallel, sometimes I found strange "Lexical error" during the analysis, and sometimes the AST is not complete.

**Comment by corradomio** [comment]
I just found a FIRST problem:

"ParserConfiguration" object doesn't contain ONLY configuration parameters, but ALSO object used to "PRE" and "POST-process" the current "provider" (the source file to analyze).

Now, because we have multiple instances of JavaParser but only a single instance of "ParserConfiguration" (there was not reason to ""configure"" each JP in different way) and then a single instance of PRE/POST processors, the result is that these processors are called by multiple threads, with different providers, with ""strange"" side effects (for example the previous ""strage"" "Lexical errors").

The most simple solution is to assign a new instance of ParserConfiguration to each JavaParser.

A more correct implementation (I think) could be to have a single ParserConfiguration  but the pre/post processors are created for each provider.

I don't know if this is enough. For now, it works !

**Comment by corradomio** [comment]
Just as a note:
analyzing sequentially 15.000 files composing "Hibernate" library I need ~4 hours.
The same code, using 7 threads, 40 minutes, with a speedup of 6x (in my log I have 6.5x).

**Comment by apixandru** [comment]
You're on to something @corradomio .

I created a similar test case where I'm sharing the parser configuration but each thread gets its own JavaParser. Digging around, I found that what's going on is that the stream is closed while others are still waiting for data.

GeneratedJavaParserTokenManager.getNextToken is not much help as it swallows exceptions without logging what's going on in 3 separate places. Seems that there are some special cases like throwing "java.io.IOException("PGCC end of stream");" but because that's way too broad, the assumption is that when we get IOException it's definitely this, but it might as well be something else. It would be great if this would be a PgccIOException instead to distinguish between flow control exceptions and actual unexpected exceptions.

Anyway, one of the culprits that appears to be attempting to use some stream after it's closed is LineEndingProcessor that wraps the original provider into it's own LineEndingProcessingProvider. 

If I disable it with setDetectOriginalLineSeparator(false) I'm no longer getting the "reading after stream was closed" error nor any other token parsing exceptions.

![image](https://user-images.githubusercontent.com/4395571/141694232-ca8e7f8a-824e-46a8-aac0-b59dfb5d3399.png)

**Comment by MysterAitch** [comment]
Thanks for looking into this @corradomio , and for sharing the workaround of supplying separate instances of the configuration -- the improvement you experienced seems to be substantial!

We're definitely open to improvements in this area of making JP more thread safe (it comes up every once in a while!), but it's not something I have any meaningful expertise in.

Similarly, @apixandru I'll need a little while to digest what you wrote but it sounds like you're onto something too.

**Comment by apixandru** [comment]
@MysterAitch similarly to @corradomio I noticed that it's not that big of a deal to just use a separate parser per thread.

My original use case was loading all compilation units in memory then linking them together to create a "find usages" index.

Since the JavaParserTypeSolver doesn't actually take a parser but a parser configuration, it was a bit of a bummer because I couldn't split the parsing into multiple threads.

Then I noticed that you can inject the symbol resolver after you parse the items so what I ended up doing was to write my own TypeSolver that rather than loading compilation units from disk would load them from memory since I already pre-loaded everything already.

My implementation of the TypeSolver is pretty much a clone of JavaParserTypeSolver, the difference being that when a type is not in the cache, instead of this.javaParser.parse(file) i'm doing a parsedFiles.get(srcFile.toAbsolutePath())

Not great but works for me.

If i use the JavaParserTypeSolver, a lot of threads will have their own copy of the compilation units and it will be uncomfortable to merge them.

**Comment by jlerbsc** [comment]
> Now, because we have multiple instances of JavaParser but only a single instance of "ParserConfiguration" (there was not reason to ""configure"" each JP in different way) and then a single instance of PRE/POST processors, the result is that these processors are called by multiple threads, with different providers, with ""strange"" side effects (for example the previous ""strage"" "Lexical errors").

From my point of view, as you noticed every instance of the ParserConfiguration class has a state which cannot be shared by different threads. Because of this you have to provide a new instance of ParserConfiguration for each parser.

**Comment by yufanyufan** [comment]
Does this mean this API is buggy, and should not be used?

```
    public SourceRoot parseParallelized(String startPackage, ParserConfiguration configuration, Callback callback)
```

**Comment by matozoid** [comment]
@jlerbsc - I think there's a design mistake (by me, I think) - ParserConfiguration is not thread safe because of two reasons: one is that you can change the configuration. That's fine, just don't change the configuration once you started. The other is that it has a few pre/postprocessor with state in them, and they get shared between thread. Woops!

My solution would be to make the user give us a factory for these processors. I would start by wrapping them in Suppliers:
```
    private final List<Supplier<PreProcessor>> preProcessors = new ArrayList<>();
    private final List<Supplier<ParseResult.PostProcessor>> postProcessors = new ArrayList<>();
```
Now for every file that gets parsed, when it is time to pre or post process, you go through the list, and add a ".get()". This will create a fresh instance for every file that gets parsed.
(working on a PR)

**Comment by yufanyufan** [comment]
Another way to solve this is to implement a clone or copy constructor for ParserConfiguration, and create a new config for each JavaParser.

**Comment by matozoid** [comment]
@yufanyufan - the answer is by the way: if you don't use the unicode preprocessor or the line ending thing, it is safe.

A copy constructor wouldn't solve it, I think, because there is no way to copy a pre or postprocessor. That's why I'm wrapping them in a Supplier factory-like thing.

**Comment by yufanyufan** [comment]
@matozoid 
Because the processor are not injected, the clone function can just create new instance of processor, instead copying them.

**Comment by MysterAitch** [comment]
I am under the impression that @matozoid 's PR in #3515 fixes the pre/post processor problems therefore this solves the issue described within the first couple of comments (this issue will auto-close once the PR checks complete and it gets merged in). 

@corradomio -- if there is any problems with this, please comment / re-open this issue (or create a new issue if you find a separate problem) :+1:
